nodestream/pipeline/pipeline.py

```diff
@@ -1,280 +1,195 @@
-        self.identifier = identifier
-        super().__init__(self.build_message())
 
-    def build_message(self):
-        message = ""
-        for section, exception in self.exceptions.items():
-            sub_message = "".join(traceback.format_tb(exception.__traceback__))
-            # Break it apart, add a tab, bring it back together.
-            message += f"\t{section}\n"
-            for sentence in [
-                "\t\t" + sentence + "\n" for sentence in sub_message.split("\n")
-            ]:
-                message += sentence
-        return f"Exceptions in {self.identifier}:\n" + message
+    __slots__ = ("step", "input", "output", "context")
 
+    def __init__(
+        self,
+        step: Step,
+        input: StepInput,
+        output: StepOuput,
+        context: StepContext,
+    ) -> None:
+        self.step = step
+        self.input = input
+        self.output = output
+        self.context = context
 
-class ForwardProgressHalted(Exception):
-    pass
+    async def start_step(self):
+        try:
+            await self.step.start(self.context)
+        except Exception as e:
+            self.context.report_error(
+                f"Error starting step {self.step.__class__.__name__}", e
+            )
 
+    async def stop_step(self):
+        try:
+            await self.step.finish(self.context)
+        except Exception as e:
+            self.context.report_error(
+                f"Error stopping step {self.step.__class__.__name__}", e
+            )
 
-class PipelineException(Exception):
-    """
-    Exception in Pipeline:
-        Output from StepException 0
-        Output from StepException 1
-    """
+    async def emit_record(self, record):
+        can_continue = await self.output.put(record)
+        if not can_continue:
+            self.context.debug(
+                "Downstream is not accepting more records. Gracefully stopping."
+            )
 
-    def __init__(self, errors: list[Exception]):
-        self.errors = errors
-        super().__init__(self.build_message())
+        return can_continue
 
-    def build_message(self):
-        message = ""
-        sub_message = ""
-        for error in self.errors:
-            sub_message += error.build_message()
+    async def drive_step(self):
+        try:
+            while (next_record := await self.input.get()) is not None:
+                results = self.step.process_record(next_record, self.context)
+                async for record in results:
+                    if not await self.emit_record(record):
+                        return
+
+            async for record in self.step.emit_outstanding_records():
+                if not await self.emit_record(record):
+                    return
+
+            self.context.debug(f"Step {self.step.__class__.__name__} finished emitting")
+        except Exception as e:
+            self.context.report_error(
+                f"Error running step {self.step.__class__.__name__}",
```

Comment:
Something like this is nice when parsing logs as messages, but it would also be nice to have the class name and index in dictionary form {step_name: "__class__.__name__", index: 1}. With something like this we wouldn't have to regex the messages to isolate something like errors by step.

Reply:
Really good call.